### PR TITLE
GH-39990: [Docs][CI] Add sphinx-lint for docs linting

### LINKED ISSUE #39990: [Docs][CI] Investigate linting C++ and other docs in CI
### Describe the enhancement requested

It would be good to look into what we currently have enabled and what we could easily add for linting docs PRs. https://github.com/apache/arrow/pull/39989 found two types of errors, both of which would be good to avoid. I think sphinx-build already offers some form of reStructuredText linting and there appear to be other tools such as https://github.com/sphinx-contrib/sphinx-lint.

- [x] Find a tool we can use for linting in developer and CI environments
- [x] Add an archery lint tasks for this
- [x] Integrate said tool into developer workflows (like pre-commit hooks)
- [ ] Integrate said tool into CI

### Component(s)

Continuous Integration, Documentation

### PR DESCRIPTION
**Pull request by amoeba**
### What changes are included in this PR?

This adds developer tooling to the repo for linting the docs by adding the sphinx-lint tool to archery and our pre-commit hooks. In both locations, only two rules are enabled at the moment (Discussed in https://github.com/apache/arrow/pull/40006): `trailing-whitespace` and `missing-final-newline`.

This PR also fixes the individual issues covered by the new tooling. 

### Are these changes tested?

Yes, though manually. I tested this works by running `archery lint --docs` and `pre-commit` without and without changes that should get caught by the rules. It works as expected.

### Are there any user-facing changes?

Yes,

1. Developers that use pre-commit hooks will see a change in behavior when they modify docs
2. Developers using archery will see a new --docs option in `archery lint`
3. Developers working on the docs may see CI failures related to the new checks

* Closes: #39990
* GitHub Issue: #39990

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (jorisvandenbossche) [inline .pre-commit-config.yaml:42]: This assumes that you as developer has installed archery? (with the other pre-commit entries here, I think pre-commit will take care of installing and caching the required tools)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

File: .pre-commit-config.yaml
Comment (amoeba) [reply .pre-commit-config.yaml:42]: Thanks for looking at this.

It does assume archery, yes. Though I'm not sure pre-commit will take care of installing/caching required tools. I'll check this though.

I opted for leaning on archery here instead of sphinx-lint directly because I felt it's simpler overall if different segments of the developer experience, including this one (docs), just require users to install archery rather than each segment requiring different tools.

Comment (amoeba) [reply .pre-commit-config.yaml]: I'm actually going to re-work this so the pre commit config sets up sphinx lint and the hook then runs sphinx-lint directly.

Comment (amoeba) [reply .pre-commit-config.yaml]: Okay this is ready for another look. I changed the pre-commit config so it doesn't need archery to be installed and instead works similarly to the other hooks like flake8 (i.e., automatically). 

One odd thing came up which is that the output gets printed twice when pre-commit runs. I did some research and it looks like pre-commit runs hooks in multiple stages and there are cases where hooks run multiple times, just in different stages. I tried manually specifying `stages`  [as outlined in the documentation](https://pre-commit.com/#confining-hooks-to-run-at-certain-stages) but that didn't seem to have any effect. Any ideas @jorisvandenbossche?

Either way, here's what this PR looks like now:

First, introduce some changes that sphinx-lint would be expected to catch and stage them:

```sh
$ git diff --staged
diff --git a/docs/source/cpp/env_vars.rst b/docs/source/cpp/env_vars.rst
index 0fa80aa11..ff2fbe52f 100644
--- a/docs/source/cpp/env_vars.rst
+++ b/docs/source/cpp/env_vars.rst
@@ -24,10 +24,12 @@ Environment Variables
 The following environment variables can be used to affect the behavior of
 Arrow C++ at runtime.  Many of these variables are inspected only once per
 process (for example, when the Arrow C++ DLL is loaded), so you cannot assume
-that changing their value later will have an effect.
+that changing their value later will have an effect.

+
 .. envvar:: ACERO_ALIGNMENT_HANDLING

+
    Arrow C++'s Acero module performs computation on streams of data.  This
    computation may involve a form of "type punning" that is technically
    undefined behavior if the underlying array is not properly aligned.  On
```

Then install archery[lint] and run:

```sh
$ archery lint --docs
INFO:archery:Running docs linter (sphinx-lint)
/Users/bryce/src/apache/arrow/docs/source/cpp/env_vars.rst:27: trailing whitespace (trailing-whitespace)
/Users/bryce/src/apache/arrow/docs/source/cpp/env_vars.rst:29: trailing whitespace (trailing-whitespace)
/Users/bryce/src/apache/arrow/docs/source/cpp/env_vars.rst:32: trailing whitespace (trailing-whitespace)
```

The output is similar if we install and run sphinx-lint directly:

```sh
$ sphinx-lint --disable all --enable trailing-whitespace,missing-final-newline docs
docs/source/cpp/env_vars.rst:27: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:29: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:32: trailing whitespace (trailing-whitespace)
```

Though we get duplicated but otherwise correct output if we run run pre-commit directly though:

```sh
$ pre-commit run
Release Audit Tool.......................................................Passed
Docker Format........................................(no files to check)Skipped
Python Format........................................(no files to check)Skipped
cython-lint..........................................(no files to check)Skipped
Sphinx Lint..............................................................Failed
- hook id: sphinx-lint
- exit code: 1

docs/source/cpp/env_vars.rst:27: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:29: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:32: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:27: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:29: trailing whitespace (trailing-whitespace)
docs/source/cpp/env_vars.rst:32: trailing whitespace (trailing-whitespace)
```

Comment (amoeba) [reply .pre-commit-config.yaml]: @jorisvandenbossche any thoughts on the above (particularly the duplicated output)?

File: .pre-commit-config.yaml
Comment (kou) [inline .pre-commit-config.yaml:61]: Can we remove `docs` from this list? (I think that duplicated logs are caused by this.)

Could you use multi-lines array style instead of one-line array style? It seems that we need to specify many command line options later. One-line array style will be difficult to understand.

```suggestion
        args:
          - --disable
          - all
          - --enable
          - trailing-whitespace,missing-final-newline
```

File: dev/archery/archery/cli.py
Comment (kou) [inline dev/archery/archery/cli.py:288]: Why do we need this?
Do we need to use custom path for `docs/`?

**Comment by kou** [comment]
Could you rebase on main?

File: .pre-commit-config.yaml
Comment (amoeba) [reply .pre-commit-config.yaml]: Nice change, thanks. Accepted.

I'm not sure about the last but but maybe we don't. Good eye.

File: dev/archery/archery/cli.py
Comment (amoeba) [reply dev/archery/archery/cli.py:288]: No, I don't think so. We do want to point `archery lint --docs` at the docs subfolder but we can hard-code that in the code. No need yet to make it user-configurable.

**Comment by amoeba** [comment]
Thanks @kou. I need to fix a couple merge conflicts during rebase and I'll ping you once this is ready.

**Comment by amoeba** [comment]
Hi @kou, I rebased and pushed a few extra changes. I just tested the behavior `archery lint --docs` and `pre-commit` and things are working as expected. This is ready for another look.

File: docs/requirements.txt
Comment (kou) [inline docs/requirements.txt:13]: Do we need this?
I think that `sphinx-lint` is installed by `pip install -e dev/archery[lint]`.

File: ci/conda_env_sphinx.txt
Comment (kou) [inline ci/conda_env_sphinx.txt:29]: Do we need this?
I think that `sphinx-lint` is installed by `pip install -e dev/archery[lint]`.

**Review by kou — APPROVED** [review]
+1

**Comment by kou** [comment]
Could you update the description before we merge this?

File: .pre-commit-config.yaml
Comment (jorisvandenbossche) [reply .pre-commit-config.yaml]: Sorry for the slow response here! 

Was the duplication fixed by the removal of the `docs` argument, or is this still something to look into? (although certainly not a blocker if it would show duplicated failures)

File: ci/conda_env_sphinx.txt
Comment (jorisvandenbossche) [reply ci/conda_env_sphinx.txt:29]: Given that archery's setup.py has its own specific version pinning, it might indeed not be needed to include here

File: docs/source/cpp/tutorials/basic_arrow.rst
Comment (jorisvandenbossche) [inline docs/source/cpp/tutorials/basic_arrow.rst:144]: This is not a correct / expected change?

Comment (amoeba) [reply docs/source/cpp/tutorials/basic_arrow.rst:144]: Thanks for catching this. When rebasing, I grabbed the wrong change. The above change undid the change in https://github.com/apache/arrow/pull/40006 so I'll push up a fix.

Comment (amoeba) [reply docs/source/cpp/tutorials/basic_arrow.rst]: Reverted in 19e01b7e5529b1a8fb56f18b9281d23c95af6cba.

File: ci/conda_env_sphinx.txt
Comment (amoeba) [reply ci/conda_env_sphinx.txt]: I agree. Removed in b526a4d844b143b6ef5c414abf0366b274645817.

File: docs/requirements.txt
Comment (amoeba) [reply docs/requirements.txt]: We don't need it, good catch. Removed in b526a4d844b143b6ef5c414abf0366b274645817.

File: .pre-commit-config.yaml
Comment (amoeba) [reply .pre-commit-config.yaml]: Yep, it goes away when I make just this one change. I'm not sure I can explain why but it's definitely the change to make.

**Comment by amoeba** [comment]
I rebased, pushed two fixes in recent changes and will merge this once CI passes.

**Comment by amoeba** [comment]
Merged. There were two failing CI checks but both were unrelated (recent pyarrow asof join failures, both have their own issues).

Thanks for the review here @kou and @jorisvandenbossche.

**Comment by jorisvandenbossche** [comment]
Thanks @amoeba! Are there other rules we want to consider enabling as follow-ups?

**Comment by amoeba** [comment]
Good reminder, thanks @jorisvandenbossche. I think quite a few would be good to add. I've filed https://github.com/apache/arrow/issues/41611 for that work.

**Pull request merged**
